### PR TITLE
(closes #1454) domain kernel ncells fix

### LINKED ISSUE #1454: [lfric] Alter the number of cells passed to 'domain' kernels
Currently PSyclone generates code that passes the number of cells to `domain` kernels as being the result of `mesh%get_ncells_2d()`. However, this causes the UM Physics code to end up looping into halos and bad things happen. We therefore need to change this to use `mesh%get_last_edge_cell()` instead.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

File: src/psyclone/tests/domain/lfric/arg_ordering_test.py [hunk 1]
```diff
@@ -106,6 +106,56 @@ def test_kernel_stub_invalid_scalar_argument():
...
+    assert create_arg_list._arglist == []
+    create_arg_list.generate()
+    assert create_arg_list._arglist == [
+        'nlayers', 'ncell_2d_no_halos', 'b', 'f1_proxy%data', 'ndf_w3',
+        'undf_w3', 'map_w3']
+
+
+def test_arg_ordering_generate_cma_kernel(dist_mem):
+    '''
+    Check that ArgOrdering generates the expected arguments for a
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/lfric/arg_ordering_test.py:134]: Oops, done.

File: doc/user_guide/dynamo0p3.rst [hunk 2]
```diff
@@ -2063,10 +2065,10 @@ identical to those for general-purpose kernels (described :ref:`above
 <dynamo0.3-stub-generation-rules>`), allowing for the fact that they
 are not permitted any type of operator argument or any argument with a
 stencil access. The only difference is that, since the kernel operates
-on the whole domain, the number of columns in the mesh (``ncell_2d``)
-must be passed in. This is provided as the second argument to the
-kernel (after ``nlayers``). ``ncell_2d`` is an ``integer`` of kind
-``i_def`` with intent ``in``.
+on the whole domain, the number of columns in the mesh, excluding those
+in the halo, (``ncell_2d_no_halos``) must be passed in. This is provided
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:2069]: Yes, I've rearranged the commas.

File: src/psyclone/tests/test_files/dynamo0p3/25.3_multikern_domain_cma.f90 [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  type(field_type) :: f1, f2, f3, f4
+  type(operator_type)            :: lma_op1
+  type(columnwise_operator_type) :: cma_op1
+
+  call invoke(                                           &
+               ! Read-write f1 (W3)
+               testkern_domain_type(b, f1),              &
+               ! Read from f1
+               testkern_type(b, f2, f3, f4, f1),         &
+               ! Write f1
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/25.3_multikern_domain_cma.f90:60]: Oops!

File: src/psyclone/tests/test_files/dynamo0p3/25.3_multikern_domain_cma.f90 [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+! POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author A. R. Porter, STFC Daresbury Lab
+
+program single_invoke
+
+  ! Description: invoke containing multiple kernels, the first 
+  ! of which is a kernel that operates on the domain instead of a
+  ! cell-column and the last of which is a CMA kernel.
+  use constants_mod,                only : r_def
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/25.3_multikern_domain_cma.f90:41]: Done.

File: src/psyclone/domain/lfric/kern_call_arg_list.py [hunk 5]
```diff
@@ -153,6 +153,20 @@ def mesh_ncell2d(self, var_accesses=None):
...
+        to the argument list and store this access in var_accesses (if
+        supplied).
+
+        :param var_accesses: optional VariablesAccessInfo instance to store \
+            the information about variable accesses.
+        :type var_accesses: \
+            :py:class:`psyclone.core.access_info.VariablesAccessInfo`
+
+        '''
+        name = self._symtab.symbol_from_tag("ncell_2d_no_halos").name
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/lfric/kern_call_arg_list.py:167]: Done.

**Comment by arporter** [comment]
CI-permitting, this is ready for review again.

**Review by TeranIvy — APPROVED** [review]
All well now. Retested with compilation, checked the modified files for `pycodestyle` and `pylint` and ran examples. The documentation builds fine and the changes are all covered.

Updated the changelog and user guide. CI permitting, this is ready to merge.

**Pull request merged**